packages/compiler-cli/ngcc/src/execution/single_process_executor.ts

```diff
@@ -10,47 +10,56 @@ import {Logger} from '../logging/logger';
+
+  doExecute(analyzeEntryPoints: AnalyzeEntryPointsFn, createCompileFn: CreateCompileFn):
+      void|Promise<void> {
+    this.logger.debug(`Running ngcc on ${this.constructor.name}.`);
+
+    const taskQueue = analyzeEntryPoints();
+    const compile =
+        createCompileFn((task, outcome) => onTaskCompleted(this.pkgJsonUpdater, task, outcome));
+
+    // Process all tasks.
+    this.logger.debug('Processing tasks...');
+    const startTime = Date.now();
+
+    while (!taskQueue.allTasksCompleted) {
+      const task = taskQueue.getNextTask() !;
+      compile(task);
+      taskQueue.markTaskCompleted(task);
+    }
+
+    const duration = Math.round((Date.now() - startTime) / 1000);
+    this.logger.debug(`Processed tasks in ${duration}s.`);
+  }
+}
 
 /**
  * An `Executor` that processes all tasks serially and completes synchronously.
  */
-export class SingleProcessExecutor implements Executor {
-  constructor(
-      private logger: Logger, private pkgJsonUpdater: PackageJsonUpdater,
-      private lockFile: LockFile) {}
-
+export class SingleProcessExecutorSync extends SingleProcessorExecutorBase implements Executor {
+  constructor(logger: Logger, pkgJsonUpdater: PackageJsonUpdater, private lockFile: SyncLocker) {
+    super(logger, pkgJsonUpdater);
+  }
   execute(analyzeEntryPoints: AnalyzeEntryPointsFn, createCompileFn: CreateCompileFn): void {
-    this.lockFile.lock(() => {
-      this.logger.debug(`Running ngcc on ${this.constructor.name}.`);
-
-      const taskQueue = analyzeEntryPoints();
-      const compile =
-          createCompileFn((task, outcome) => onTaskCompleted(this.pkgJsonUpdater, task, outcome));
-
-      // Process all tasks.
-      this.logger.debug('Processing tasks...');
-      const startTime = Date.now();
-
-      while (!taskQueue.allTasksCompleted) {
-        const task = taskQueue.getNextTask() !;
-        compile(task);
-        taskQueue.markTaskCompleted(task);
-      }
-
-      const duration = Math.round((Date.now() - startTime) / 1000);
-      this.logger.debug(`Processed tasks in ${duration}s.`);
-    });
+    this.lockFile.lock(() => this.doExecute(analyzeEntryPoints, createCompileFn));
   }
 }
 
 /**
  * An `Executor` that processes all tasks serially, but still completes asynchronously.
  */
-export class AsyncSingleProcessExecutor extends SingleProcessExecutor {
-  async execute(...args: Parameters<Executor['execute']>): Promise<void> {
-    return super.execute(...args);
+export class SingleProcessExecutorAsync extends SingleProcessorExecutorBase implements Executor {
+  constructor(logger: Logger, pkgJsonUpdater: PackageJsonUpdater, private lockFile: SyncLocker) {
+    super(logger, pkgJsonUpdater);
+  }
+  async execute(analyzeEntryPoints: AnalyzeEntryPointsFn, createCompileFn: CreateCompileFn):
+      Promise<void> {
+    await this.lockFile.lock(async() => this.doExecute(analyzeEntryPoints, createCompileFn));
```

Comment:
I think it's strange to pass an async fn to the synchronous locker, given that the `SyncLocker` does not await its result before unlocking the lock file. That's actually fine because all work still executes synchronously, but I think it should be enough to have just the `async` marker on the `execute` method as that ought to be sufficient to achieve asynchronous completion.

```suggestion
    this.lockFile.lock(() => this.doExecute(analyzeEntryPoints, createCompileFn));
```

Reply:
Actually it seems that these are actually needed. CI errors:

```
ERROR: /home/circleci/ng/packages/compiler-cli/ngcc/BUILD.bazel:5:1: Compiling TypeScript (devmode) //packages/compiler-cli/ngcc:ngcc failed (Exit 1)
packages/compiler-cli/ngcc/src/execution/single_process_executor.ts:63:5 - error TS21225: [tsetse] All Promises in async functions must either be awaited or used in an expression.
	See http://tsetse.info/must-use-promises

63     this.lockFile.lock(() => this.doExecute(analyzeEntryPoints, createCompileFn));
```

Reply:
I guess that is because `doExecute` has a return type of `void|Promise<void>`, whereas it never returns a Promise in reality (that would be an issue for the sync variant). If that return type was changed into just `void`, the `SyncLocker.lock` call would also return `void` so tsetse would no longer flag it as needing to be awaited.